Wizard: Add group value on blur, Add user, and Next actions

mgold1234 opened the pull request:
This commit fixes a bug where group field value was only added to the list on Enter key press. Now the value is added to the list also when:
- Blurring the field (onBlur)
- Clicking "Add user" button
- Clicking "Next" button

Fixes: https://github.com/osbuild/image-builder-frontend/issues/3943

regexowl commented:
Maybe we could add some kind of guidance or a helper text to let people know that they can add a group on Enter keydown? To let people know about things without doing changes they might not want. I don't know, just more thoughts...

regexowl commented on src/Components/CreateImageWizard/LabelInput.tsx:
```diff
@@ -63,12 +63,17 @@ const LabelInput = ({
 
   const addItem = (value: string) => {
-    if (list?.includes(value) || requiredList?.includes(value)) {
+    const trimmedValue = value.trim();
```
Nice! :heart:

regexowl reviewed (changes requested):
Thanks for double checking the behaviour with the UX team! :heart:

The on blur and on add a user behaviour looks great. Let's just tweak the Next button behaviour: _"trigger validation when "Next" is clicked and if there are any errors to render them and block the Next button"_

regexowl commented on src/Components/CreateImageWizard/steps/Users/components/UserInfo.tsx:
```diff
@@ -8,8 +8,38 @@ import { useAppSelector } from '../../../../../store/hooks';
-const UserInfo = () => {
+type UserInfoProps = {
+  usersPristine: boolean;
+  onInputErrorsChange?: ((hasAnyError: boolean) => void) | undefined;
```
I'm wondering if the `onInputErrorsChange` could also be removed from the passed props to make things simpler. Looking at the filesystem again, the difference between that validation and users would be that this works with an input directly, meaning outside of state.

Just thinking out loud - would it make sense to create an optional value for group input in the store so it can easily get to the validator? It would need to be handled in the store, but it might mean less `useEffect`s and keeping up with refs. What do you think?

mgold1234 committed:
Wizard: Add group value on blur, Add user, and Next actions

This commit fixes a bug where group field value was only added to the list on Enter key press.
Now the value is added to the list also when:
- Blurring the field (onBlur)
- Clicking "Add user" button
- Clicking "Next" button

Store pending group input in Redux (pendingGroupInput) and commit it
explicitly via commitAllPendingGroupInputs on Enter, Add User, and
Next. Use relatedTarget in handleBlur to skip addItem when focus
moves to a button, avoiding DOM shifts that cause double-click issues.

regexowl commented on src/Components/CreateImageWizard/steps/Users/utilities/isUserDefined.ts:
```diff
@@ -0,0 +1,18 @@
+ * - SSH key
+ * - groups
+ * - current group input (typing in progress)
+ */
```
Unnecessary comment, the info is evident from the function itself.

regexowl reviewed (changes requested):
Ah, I see. I've started reviewing, but there seems to be older users step than we have currently. It's really weird that the PR doesn't warn about conflicts, but there was a bigger change to the way the users table is rendered.

Before (and what I see in this PR):
<img width="1344" height="392" alt="image" src="https://github.com/user-attachments/assets/45ee6b98-81ab-4c1e-ae67-256d73a78b55" />

And after (what is currently on stage after SPUR changes):
<img width="1396" height="382" alt="image" src="https://github.com/user-attachments/assets/87ffe901-f7e5-4919-8079-98913a2dd4df" />

Could you please rebase so the code is up-to-date with main? I'll continue with the review after the rebase.

regexowl commented:
Noticed some weird behaviour. When there's more groups the "Add user" and "Next" button don't "work" any more unless the new group is added first. So realistically - the user would have to click the Next or Add button twice to do the action they wanted to do:

[Screencast_20260212_100920.webm](https://github.com/user-attachments/assets/b691514f-67e4-4a34-96d9-9fdebb8176fa)

Can you also please squash the commits so the `.gitignore` changes cancel each other?

lucasgarfield commented on src/Components/CreateImageWizard/CreateImageWizard.tsx:
```diff
@@ -125,8 +125,12 @@ export const CustomWizardFooter = ({
+            }
             if (!beforeNext || beforeNext()) goToNextStep();
           }}
+          onMouseDown={(e) => e.preventDefault()}
```
`preventDefault` on `mousedown` suppresses native focus management. This breaks keyboard navigation — users tabbing to this button won't get expected focus behavior, and screen readers may not announce it correctly.

This exists to prevent the browser's natural blur from firing before `onClick`, but it's treating a symptom. If input commitment moves to `beforeNext`, this workaround is unnecessary. Same issue on the "Review and finish" button at line 163.

lucasgarfield commented on src/Components/CreateImageWizard/CreateImageWizard.tsx:
```diff
@@ -527,11 +532,10 @@ const CreateImageWizard = ({ isEdit }: CreateImageWizardProps) => {
-                  userGroupsValidation.disabledNext
-                  ? 'error'
-                  : 'default'
+              usersStepAttemptedNext &&
```
This changes when the nav item shows error status:
- **Before:** after *visiting* the step (`wasUsersVisited`)
- **After:** after *attempting to proceed* (`usersStepAttemptedNext`)

This isn't mentioned in the PR description. It also creates an inconsistency — the Registration step still uses `wasRegisterVisited` for the same purpose.

If intentional, call it out in the description and consider aligning Registration to the same pattern.

lucasgarfield commented on src/Components/CreateImageWizard/LabelInput.tsx:
```diff
@@ -46,33 +50,58 @@ const LabelInput = ({
   const dispatch = useAppDispatch();
 
-  const [inputValue, setInputValue] = useState('');
+  const useReduxState =
```
Two problems:

1. `useReduxState` — the `use` prefix is reserved for hooks. This is a boolean. `isReduxControlled` or similar.

2. This component now switches between controlled (Redux) and uncontrolled (local state) modes at runtime based on prop presence. That creates two divergent code paths and requires non-null assertions (`currentInputValue!` on line 63, `onInputValueChange!` on lines 85 and 152) to satisfy TypeScript.

Fix with discriminated union types:
```tsx
type ControlledProps = {
  currentInputValue: string;
  onInputValueChange: (value: string) => UnknownAction;
};
type UncontrolledProps = {
  currentInputValue?: never;
  onInputValueChange?: never;
};
type LabelInputProps = (ControlledProps | UncontrolledProps) & { /* shared */ };
```
No runtime checks, no `!` assertions — TypeScript enforces that both are present or both absent.

lucasgarfield commented on src/Components/CreateImageWizard/LabelInput.tsx:
```diff
@@ -46,33 +50,58 @@ const LabelInput = ({
-      stepValidation.errors[fieldName].split('|');
-  }
+  const validationErrorText = stepValidation.errors[fieldName] || '';
+  const validationErrors = React.useMemo(
```
This `useMemo` is splitting a short error string — the memoization overhead likely exceeds the computation cost. Compute inline.

Also, line 67 uses `||` for the default value. Prefer `??` — `||` coerces falsy values like `0` and `''`, which can mask bugs in other contexts. Consistency matters.

The pipe-delimited error format itself is fragile (if an error message ever contains `|`, parsing breaks), but that's pre-existing.

lucasgarfield commented on src/Components/CreateImageWizard/LabelInput.tsx:
```diff
@@ -46,33 +50,58 @@ const LabelInput = ({
+    [validationErrorText],
+  );
+
+  useEffect(() => {
```
This `useEffect` syncs derived error state to the parent via callback. Two problems:

1. If `onInputErrorChange` isn't memoized with `useCallback`, this effect fires every render. If the parent updates state in response, you get an infinite loop.

2. The parent's error view lags one render behind reality (`useEffect` fires after render).

Better approach: if validation state is in Redux, both parent and child read from Redux directly — no callback, no sync lag. `hasError` is derived, not synchronized.

lucasgarfield commented on src/Components/CreateImageWizard/LabelInput.tsx:
```diff
@@ -46,33 +50,58 @@ const LabelInput = ({
 
-  const addItem = (value: string) => {
-    if (list?.includes(value) || requiredList?.includes(value)) {
+  const addItem = (value: string): boolean => {
```
`addItem` validates, sets error state, dispatches to Redux, calls parent callbacks, and returns a boolean. The return value is unused at every call site.

Separate validation from action:
```tsx
const validateItem = (value: string): string | null => {
  const trimmed = value.trim();
  if (!trimmed) return null;
  if (list?.includes(trimmed)) return `${item} already exists.`;
  if (!validator(trimmed)) return getErrorMessage(fieldName);
  return null;
};

const addItem = (value: string) => {
  const error = validateItem(value);
  if (error) {
    setOnStepInputErrorText(error);
    return;
  }
  dispatch(addAction(value.trim()));
  setOnStepInputErrorText('');
};
```
Validation is now testable in isolation. `addItem` does one thing.

lucasgarfield commented on src/Components/CreateImageWizard/LabelInput.tsx:
```diff
@@ -130,14 +166,30 @@ const LabelInput = ({
-  if (invalidImports) errors.push(invalidImports);
-  if (duplicateImports) errors.push(duplicateImports);
+  // Filter out validation errors that are already shown in onStepInputErrorText
+  const filteredValidationErrors = validationErrors.filter((error) => {
```
Deduplicating errors via `includes('already exists')` / `includes('Group already exists')` — this will break silently if anyone changes the error wording, and it only handles one specific case.

If both `onStepInputErrorText` and `stepValidation.errors` can produce the same error, the error sources aren't coordinated. Either use error codes for identity, or have a single source of truth so deduplication isn't needed.

lucasgarfield reviewed (changes requested):
The bug fix is needed — group values shouldn't be silently dropped on Next. But the approach has problems.

The core issue: this PR coordinates input commitment through DOM events (`document.activeElement.blur()` + `onMouseDown preventDefault`) instead of explicit state management. This creates a race condition where `beforeNext` reads stale validation state, and breaks keyboard accessibility.

**What needs to change:**
- Replace DOM manipulation with explicit input commitment in `beforeNext`
- Remove `onMouseDown preventDefault` (accessibility regression)
- Fix the controlled/uncontrolled mode switching in `LabelInput`
- Remove `useEffect` that syncs derived error state to parent
- Add tests for the new blur behavior

The `??` refactoring and input trimming are fine. Details inline.

regexowl reviewed (changes requested):
Ran through the code and added some comments.

Also noticed two possible bugs:
- Invalid groups get silently dropped on "Add user" and "Next"
- Duplicate groups get silently dropped on "Add user" and "Next"
